### DISCUSSION TITLE
Equivalent of SKCanvasView

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#12269] [Q&A] [answered]
What Avalonia control could be equivalent of [SKCanvasView](https://learn.microsoft.com/en-us/dotnet/api/skiasharp.views.forms.skcanvasview?view=skiasharp-views-forms-2.88&devlangs=csharp&f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(SkiaSharp.Views.Forms.SKCanvasView)%3Bk(DevLang-csharp)%26rd%3Dtrue)?

**Comment by robloo** [comment]
There isn't one that I can find.

On my end, for the purpose of getting WinUI code running quickly, the WinUI version of this control was roughly ported. Its more of a prototype until a better solution is found.

Long term, like with WPF, things should render directly inside the control. You can still grab the skia canvas in the render method if you have to. That is another interrum solution.

Using the Render override will then eliminate the need for an SKCanvasView because relevant controls will just be rendering themselves with Avalonia's low level API. That wasn't really possible with UWP/WinUI.

Bottom line a quick and dirty solution was done until SKCanvasView can be removed and controls just render themselves using Avalonia's abstracted API. If impeller becomes the new standard (SkiaSharp is hardly updated these days) that will eliminate the need for big rewrites.

**Reply by workgroupengineering** [reply]
thanks, i try to solve using ICustomDrawOperation.


a question.  Is SKCanvas unique for a TopLevel?  

When I call SKCanvas.Clear it clears all instances of the control present in the TopLevel.

Is it a bug of SKCanvas.Clear
that doesn't respect the Clip?

**Reply by maxkatz6** [reply]
This is a direct canvas of the window (render target), so this behavior is expected.

If you want behavior similar to SKCanvasView in other frameworks, you need to create your own SKCanvas, and then draw it on top of the avalonia one. Obviously, it's suboptimal but might be required for some tasks.

**Reply by robloo** [reply, marked as answer]
@workgroupengineering Here is the code for the control on my end (you'll need to copy the SKPainSurfaceEventArgs too) https://github.com/mono/SkiaSharp/blob/main/source/SkiaSharp.Views/SkiaSharp.Views.Shared/SKPaintSurfaceEventArgs.cs. As @maxkatz6 stated above it creates its own SKCanvas.

Like I said, it's pretty messy and just hacked together to get things to run for now during the port. It isn't intended for long-term use but you are welcome to try this code under any license you wish (original MIT). If it's useful maybe it can be used as a base to get a control into SkiaSharp or even in Avalonia's control labs repo.

```csharp
using Avalonia;
using Avalonia.Controls;
using Avalonia.Interactivity;
using Avalonia.Layout;
using Avalonia.Media;
using Avalonia.Media.Imaging;
using Avalonia.Platform;
using Avalonia.Skia;
using Avalonia.Threading;
using SkiaSharp;
using System;

namespace Avalonia.Controls
{
    /// <summary>
    /// A Xaml canvas control that can be drawn on using SkiaSharp drawing commands.
    /// </summary>
    /// <remarks>
    /// See: https://github.com/mono/SkiaSharp/blob/main/source/SkiaSharp.Views/SkiaSharp.Views.UWP/SKXamlCanvas.cs.
    /// </remarks>
    public partial class SKXamlCanvas : Canvas
    {
        /// <summary>
        /// Event to externally paint the Skia surface (using the <see cref="SKCanvas"/>).
        /// </summary>
        public event EventHandler<SKPaintSurfaceEventArgs> PaintSurface;

        private static readonly Vector Dpi = new Vector(96, 96);

        private bool _IgnorePixelScaling;

        /// <summary>
        /// Initializes a new instance of the <see cref="SKXamlCanvas"/> class.
        /// </summary>
        public SKXamlCanvas()
        {
        }

        /// <summary>
        /// Gets the current pixel size of the canvas.
        /// Any scaling factor is already applied.
        /// </summary>
        public Size CanvasSize { get; private set; }

        /// <summary>
        /// Gets the current render scaling applied to the control.
        /// </summary>
        public double Scale { get; private set; } = 1;

        /// <summary>
        /// Gets or sets a value indicating whether the canvas's resolution and scale
        /// will be automatically adjusted to match physical device pixels.
        /// </summary>
        public bool IgnorePixelScaling
        {
            get => this._IgnorePixelScaling;
            set
            {
                this._IgnorePixelScaling = value;
                this.Invalidate();
            }
        }

        /*
        private void OnDpiChanged(DisplayInformation sender, object args = null)
        {
            Dpi = sender.LogicalDpi / DpiBase;
            Invalidate();
        }
        */

        /// <summary>
        /// Invalidates the canvas causing the surface to be repainted.
        /// This will fire the <see cref="PaintSurface"/> event.
        /// </summary>
        public void Invalidate()
        {
            Dispatcher.UIThread.Post(() => this.RepaintSurface());
        }

        /// <summary>
        /// Repaints the Skia surface and canvas.
        /// </summary>
        private void RepaintSurface()
        {
            if (!this.IsVisible)
            {
                return;
            }

            // Display scaling is important to consider here:
            // The bitmap itself should be sized to match physical device pixels.
            // This ensures it is never pixelated and renders properly to the display.
            // However, in several cases physical pixels do not match the logical pixels.
            // We also don't want to have to consider scaling in external code when calculating graphics.
            // To make this easiest, the layout scaling factor is calculated and then used
            // to find the size of the bitmap. This ensures it will match device pixels.
            // Then the canvas undoes this by setting a scale factor itself.
            // This means external code can use logical pixel size and the canvas will transform as needed.
            // Then the underlying bitmap is still at physical device pixel resolution.

            if (this.IgnorePixelScaling)
            {
                this.Scale = 1;
            }
            else
            {
                this.Scale = LayoutHelper.GetLayoutScale(this);
            }

            int pixelWidth = Convert.ToInt32(this.Bounds.Width * this.Scale);
            int pixelHeight = Convert.ToInt32(this.Bounds.Height * this.Scale);
            this.CanvasSize = new Size(pixelWidth, pixelHeight);

            // WriteableBitmap does not support zero-size dimensions
            // Therefore, to avoid a crash, exit here if size is zero
            if (pixelWidth == 0 ||
                pixelHeight == 0)
            {
                this.Background = null;
                return;
            }

            var bitmap = new WriteableBitmap(
                new PixelSize(pixelWidth, pixelHeight),
                Dpi,
                PixelFormat.Bgra8888,
                AlphaFormat.Premul);

            using (var framebuffer = bitmap.Lock())
            {
                var info = new SKImageInfo(
                    framebuffer.Size.Width,
                    framebuffer.Size.Height,
                    framebuffer.Format.ToSkColorType(),
                    SKAlphaType.Premul);

                var properties = new SKSurfaceProperties(SKPixelGeometry.RgbHorizontal);

                // It is not too expensive to re-create the SKSurface on each re-paint.
                // See: https://groups.google.com/g/skia-discuss/c/3c10MvyaSug/m/UOr238asCgAJ
                //
                // When creating the SKSurface it is important to specify a pixel geometry
                // A defined pixel geometry is required for some anti-aliasing algorithms such as ClearType
                // Also see: https://github.com/AvaloniaUI/Avalonia/pull/9558
                using (var surface = SKSurface.Create(info, framebuffer.Address, framebuffer.RowBytes, properties))
                {
                    if (!this.IgnorePixelScaling)
                    {
                        surface.Canvas.Scale(Convert.ToSingle(this.Scale));
                    }

                    this.OnPaintSurface(new SKPaintSurfaceEventArgs(surface, info, info));
                }

                properties.Dispose();
            }

            this.Background = new ImageBrush(bitmap)
            {
                AlignmentX = AlignmentX.Left,
                AlignmentY = AlignmentY.Top,
                Stretch = Stretch.Fill
            };

            return;
        }

        /// <inheritdoc/>
        protected override void OnLoaded(RoutedEventArgs e)
        {
            base.OnLoaded(e);

            /*
            var display = DisplayInformation.GetForCurrentView();
            display.DpiChanged += OnDpiChanged;

            OnDpiChanged(display);
            */
        }

        /// <inheritdoc/>
        protected override void OnUnloaded(RoutedEventArgs e)
        {
            base.OnUnloaded(e);

            /*
            var display = DisplayInformation.GetForCurrentView();
            display.DpiChanged -= OnDpiChanged;
            */
        }

        /// <inheritdoc/>
        protected override void OnSizeChanged(SizeChangedEventArgs e)
        {
            base.OnSizeChanged(e);

            this.Invalidate();
            return;
        }

        /// <summary>
        /// Called when the canvas should repaint its surface.
        /// </summary>
        /// <param name="e">The event args.</param>
        protected virtual void OnPaintSurface(SKPaintSurfaceEventArgs e)
        {
            this.PaintSurface?.Invoke(this, e);
        }

        /// <inheritdoc/>
        protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change)
        {
            base.OnPropertyChanged(change);

            if (change.Property == IsVisibleProperty)
            {
                this.Invalidate();
            }

            return;
        }
    }
}

```
